### ISSUE TITLE
angled overhangs not detected?

### ISSUE THREAD

**Issue by halfshavedyaks** [smartavionics/Cura#125]
**Application version**
Cura-mb-master-x86_64-20210629.AppImage

**Platform**
linux Mint 20

**Printer**
modified CR10S

**Reproduction steps**
set overhang speed (70%) and overhang angle (20deg) and slice on a model with overhangs at various XY angles

**Screenshot(s)**
[overhang-proj&screen.zip](https://github.com/smartavionics/Cura/files/7111430/overhang-proj.screen.zip)

**Actual results**
angled overhang isn't detected

**Expected results**
print speed should slow down on angled overhang same as on orthogonal overhang

**Project file**
included in above zip file

**Log file**
is the log needed? I'll find it if requested

**Comment by halfshavedyaks** [comment]
![overhang](https://user-images.githubusercontent.com/33595281/132124137-0bdfa2f9-e19f-4e29-9c12-7a97fa07ba68.jpeg)
added image separately for convenience

**Comment by smartavionics** [comment]
Hmm, that's interesting, I'll investigate...

**Comment by smartavionics** [comment]
I've come up with a fix and will test it some more before my next release. Thanks for posting the issue!

![Screenshot_2021-09-05_12-09-25](https://user-images.githubusercontent.com/585618/132124633-c65d6a6b-2ffd-41c9-b31c-b9b79ad2770d.png)

**Comment by halfshavedyaks** [comment]
that was fast - that's great!

is there a way I can test it without having to build cura myself?

**Comment by smartavionics** [comment]
I would need to build a new AppImage for you to test it and I will be doing that when the upstream 4.11 release arrives which should be coming along in a few days I guess.

**Comment by halfshavedyaks** [comment]
sounds great! Will this fix apply to the overhang fan settings too?

**Comment by smartavionics** [comment]
Yes, the problem was that it wasn't recognising those lines as being overhung at all but now it will be so both the speed and the fan will be altered for them.

**Comment by halfshavedyaks** [comment]
I've been testing and I see that the existing release does work on most overhangs, but has trouble with some. I found another case with weird behaviour in the current release that might be worth checking if your fix covers it.

![overhangborkage](https://user-images.githubusercontent.com/33595281/132190179-5d657cca-9d9c-4759-a0ff-7f5c907c89a8.jpeg)

[overhangborkage.zip](https://github.com/smartavionics/Cura/files/7114661/overhangborkage.zip)

incidentally and OT that project also seems to show an apparent bug where 3 skirt lines are printed even though only one is set, but I'm very new to cura so maybe I missed something.

**Comment by halfshavedyaks** [comment]
I've been looking at the g-code files in prusa g-code viewer, and the fan speed increases are not happening for the same lines that the print speed decreases are, they are being triggered separately, and they are wrong in different ways from each other.

fan speed is being increased for adjacent lines and previous/subsequent lines where it should not be.

I'll post screenshots and files when I get time. but this is just a alert if debugging this feature to check both fan and print speed since they are not reacting the same way to overhangs.

EDIT - false alarm, I now think I'm quite wrong about this, at least in some cases the extra lines with high fan speed are caused by the fan speed lag setting - I guess there's no way to turn on the fan in the middle of a line, so it takes the whole line needed to achieve the fan lag amount.

**Comment by smartavionics** [comment]
I'm not at home now to check but remember that the fan can't speed up instantly so the speed increase will be commanded some distance before the overhang is printed.MarkOn 6 Sep 2021 14:43, halfshavedyaks ***@***.***> wrote:
I've been looking at the g-code files in prusa g-code viewer, and the fan speed increases are not happening for the same lines that the print speed decreases are, they are being triggered separately, and they are wrong in different ways from each other.
fan speed is being increased for adjacent lines and previous/subsequent lines where it should not be.
I'll post screenshots and files when I get time. but this is just a alert if debugging this feature to check both fan and print speed since they are not reacting the same way to overhangs.

**Comment by halfshavedyaks** [comment]
yes you are right, as per my edit, sorry if you didn't see that before replying.

The fan will turn on early, but it can't turn on mid line (I presume) so it will turn on at the start of the line that is closest to the overhang plus the spinup time... so the amount of the print that looks like it has faster fan varies depending on the line length. It was bit confusing to see until I realised what was happening.

**Comment by halfshavedyaks** [comment]
On closer examination I'm still seeing some strange fan behaviour that isn't just the spinup time. Here's a project that shows odd fan behaviour on the bridge, the fan stops early on the bridge (is there spindown time too?) and it also doesn't spinup in advance for some of the perimeter bridge sections.

There's also a small area where overhangs are not detected properly, and some fan issues there too, although maybe that overhang area will be fixed with your fix.

![fansspeed](https://user-images.githubusercontent.com/33595281/132374231-844bc7bb-ec65-4bbe-8d05-388ca6db1bd1.png)

[oh&bridgetest.zip](https://github.com/smartavionics/Cura/files/7122785/oh.bridgetest.zip)

**Comment by smartavionics** [comment]
> On closer examination I'm still seeing some strange fan behaviour that isn't just the spinup time. Here's a project that shows odd fan behaviour on the bridge, the fan stops early on the bridge 

> (is there spindown time too?) 

yes

> and it also doesn't spinup in advance for some of the perimeter bridge sections.

That's true.

> There's also a small area where overhangs are not detected properly, and some fan issues there too, although maybe that overhang area will be fixed with your fix.

yes

**Comment by halfshavedyaks** [comment]
I'm not seeing spindown applied to perimeters anywhere. I'm not sure it bothers me but it seems inconsistent.

also I think spinup length (which is always whole lines) may be being calculated using the overhang printing speed, but since it happens before the overhang begins, it should actually be calculated on the speed of the preceding perimeter(s)

my cooling fan won't start from stopped if the pwm amount is below about 40%. I suggest adding a feature like in s3d, that blips the fan to 100% briefly when starting from stopped. S3D pauses for 500ms to do this, which i think is excessive, I suspect a 100ms shorter pause would do, or better still to do it while still printing.

**Comment by halfshavedyaks** [comment]
I've been using the new release and it is much better thanks!

There are still some small areas where overhang is not detected correctly, but they are ignorable, I haven't found anything likely to cause a significant problem with the print yet.
